ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -938,6 +938,16 @@ static bool ShouldHandleFileOpEvent(
         }
     }
     
+    {
+        PerfSample checkRootOnlineSample(perfTracer, PrjFSPerfCounter_FileOp_ShouldHandle_CheckRootOnline);
+        
+        if (!VirtualizationRoot_IsOnline(*root))
```

Comment:
This seems like similar to the check done in HandleVnodeOperation -> TryGetVirtualizationRoot -> !VirtualizationRoot_IsOnline
HandleFileOpOperation -> ShouldHandleFileOpEvent -> !VirtualizationRoot_IsOnline

In ShouldHandleFileOpEvent, we do a VirtualizationRoot_IsValidRootHandle check but not in TryGetVirtualizationRoot.  Can you elaborate on why it's needed in one and not the othere?

Reply:
I guess `TryGetVirtualizationRoot()` checks the 2 distinct cases of invalid handles, `RootHandle_ProviderTemporaryDirectory` and `RootHandle_None` separately.

I generally have a bunch of issues with this whole area of code:

 * `TryGetVirtualizationRoot` is terribly named for such a "kitchen sink" function.
 * There's a lot of duplication between the fileop and vnode listener checks, but no code reuse.
 * The functions take far too many parameters.

These aren't issues I'm planning to fix in this particular PR; I'd love to fix them once the fileop and vnode handlers are unit tested though.

Reply:
@pmj Yes Yes Yes :-)

> There's a lot of duplication between the fileop and vnode listener checks, but no code reuse.
Exactly and as you called out the checks are different, probably contributing to this bug happening.

> These aren't issues I'm planning to fix in this particular PR; I'd love to fix them once the fileop and vnode handlers are unit tested though.
Let's make that a plan!

Reply:
> Exactly and as you called out the checks are different, probably contributing to this bug happening.

I agree on this point as well, if we can get those code paths sharing code again that would be great 😄